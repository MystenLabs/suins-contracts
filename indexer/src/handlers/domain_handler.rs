use crate::models::sui::dynamic_field::Field;
use crate::models::suins::domain::Domain;
use crate::models::suins::name_record::NameRecord;
use crate::models::suins::subdomain_registration::SubDomainRegistration;
use crate::models::{NameRecordChange, SuinsCheckpointData, SuinsIndexerCheckpoint};
use anyhow::anyhow;
use async_trait::async_trait;
use diesel::dsl::case_when;
use diesel::upsert::excluded;
use diesel::{BoolExpressionMethods, ExpressionMethods};
use diesel_async::scoped_futures::ScopedFutureExt;
use diesel_async::{AsyncConnection, RunQueryDsl};
use futures::future::try_join_all;
use move_core_types::language_storage::StructTag;
use std::collections::HashSet;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::{Connection, Db};
use sui_types::base_types::SuiAddress;
use sui_types::full_checkpoint_content::{CheckpointData, CheckpointTransaction};
use sui_types::object::Object;

#[macro_export]
macro_rules! update_field_query {
    ($field:ident) => {{
        case_when(
            excluded(last_checkpoint_updated).gt(last_checkpoint_updated),
            excluded($field),
        )
        .otherwise($field)
    }};
}

pub struct DomainHandler {
    registry_table_id: SuiAddress,
    subdomain_wrapper_type: StructTag,
    name_record_type: StructTag,
}

impl DomainHandler {
    pub fn new(
        registry_table_id: SuiAddress,
        subdomain_wrapper_type: StructTag,
        name_record_type: StructTag,
    ) -> Self {
        Self {
            registry_table_id,
            name_record_type,
            subdomain_wrapper_type,
        }
    }

    /// Filter by the dynamic field value type.
    /// A valid name record for an object has the type `Field<Domain,NameRecord>,
    /// and the parent of it is the `registry` table id.
    pub fn is_name_record(&self, object: &Object) -> bool {
        object
            .get_single_owner()
            .is_some_and(|owner| owner == self.registry_table_id)
            && object
                .struct_tag()
                .is_some_and(|tag| tag == self.name_record_type)
    }

    /// Checks if the object referenced is a subdomain wrapper.
    /// For subdomain wrappers, we're saving the ID of the wrapper object,
    /// to make it easy to locate the NFT (since the base NFT gets wrapped and indexing won't work there).
    pub fn is_subdomain_wrapper(&self, object: &Object) -> bool {
        object
            .struct_tag()
            .is_some_and(|tag| tag == self.subdomain_wrapper_type)
    }

    /// Parses the name record changes + subdomain wraps.
    /// and pushes them into the supplied vector + hashmap.
    ///
    /// It is implemented in a way to do just a single iteration over the objects.
    pub fn parse_record_changes(
        &self,
        results: &mut SuinsIndexerCheckpoint,
        objects: &[Object],
    ) -> anyhow::Result<()> {
        for object in objects {
            // Parse all the changes to a `NameRecord`
            if self.is_name_record(object) {
                let name_record: Field<Domain, NameRecord> = object
                    .to_rust()
                    .ok_or_else(|| anyhow!("Failed to parse name record for {:?}", object))?;

                let id = object.id();

                // Remove from the removals list if it's there.
                // The reason it might have been there is that the same name record might have been
                // deleted in a previous transaction in the same checkpoint, and now it got re-created.
                results.removals.remove(&id);

                results
                    .name_records
                    .insert(id, NameRecordChange(name_record));
            }
            // Parse subdomain wrappers and save them in our hashmap.
            // Later, we'll save the id of the wrapper in the name record.
            // NameRecords & their equivalent SubdomainWrappers are always created in the same PTB, so we can safely assume
            // that the wrapper will be created on the same checkpoint as the name record and vice versa.
            if self.is_subdomain_wrapper(object) {
                let sub_domain: SubDomainRegistration = object.to_rust().ok_or_else(|| {
                    anyhow!(
                        "Failed to deserialize SubDomainRegistration object {:?}",
                        object
                    )
                })?;
                results
                    .subdomain_wrappers
                    .insert(sub_domain.nft.domain_name, sub_domain.id.to_string());
            };
        }
        Ok(())
    }

    /// Parses a list of the deletions in the checkpoint and adds them to the removals list.
    /// Also removes any name records from the updates, if they ended up being deleted in the same checkpoint.
    pub fn parse_record_deletions(
        &self,
        results: &mut SuinsIndexerCheckpoint,
        transaction: &CheckpointTransaction,
    ) {
        // a list of all the deleted objects in the transaction.
        let deleted_objects: HashSet<_> = transaction
            .effects
            .all_tombstones()
            .into_iter()
            .map(|(id, _)| id)
            .collect();

        for input in transaction.input_objects.iter() {
            if self.is_name_record(input) && deleted_objects.contains(&input.id()) {
                // since this record was deleted, we need to remove it from the name_records hashmap.
                // that catches a case where a name record was edited on a previous transaction in the checkpoint
                // and deleted from a different tx later in the checkpoint.
                results.name_records.remove(&input.id());

                // add it in the list of removals
                results.removals.insert(input.id());
            }
        }
    }
}

#[async_trait]
impl Handler for DomainHandler {
    type Store = Db;

    async fn commit<'a>(
        values: &[Self::Value],
        conn: &mut Connection<'a>,
    ) -> anyhow::Result<usize> {
        use crate::schema::domains::*;
        // split data
        let (updates, removals) =
            values
                .iter()
                .fold((vec![], vec![]), |(mut updates, mut removals), value| {
                    updates.extend(value.updates.clone());
                    removals.push((value.checkpoint, value.removals.clone()));
                    (updates, removals)
                });

        Ok(conn
            .transaction(|conn| {
                async move {
                    if updates.is_empty() && removals.is_empty() {
                        return Ok::<_, anyhow::Error>(0);
                    }
                    // commit all Verified Domains
                    let mut changes = 0usize;
                    if !updates.is_empty() {
                        // Bulk insert all updates and override with data.
                        changes += diesel::insert_into(table)
                            .values(updates)
                            .on_conflict(name)
                            .do_update()
                            .set((
                                expiration_timestamp_ms
                                    .eq(update_field_query!(expiration_timestamp_ms)),
                                nft_id.eq(update_field_query!(nft_id)),
                                target_address.eq(update_field_query!(target_address)),
                                data.eq(update_field_query!(data)),
                                last_checkpoint_updated
                                    .eq(update_field_query!(last_checkpoint_updated)),
                                field_id.eq(update_field_query!(field_id)),
                                // We always want to respect the subdomain_wrapper re-assignment, even if the checkpoint is older.
                                // That prevents a scenario where we first process a later checkpoint that did an update to the name record (e..g change target address),
                                // without first executing the checkpoint that created the subdomain wrapper.
                                // Since wrapper re-assignment can only happen every 2 days, we can't write invalid data here.
                                subdomain_wrapper_id.eq(case_when(
                                    excluded(subdomain_wrapper_id).is_not_null(),
                                    excluded(subdomain_wrapper_id),
                                )
                                .otherwise(subdomain_wrapper_id)),
                            ))
                            .execute(conn)
                            .await?;
                    }

                    // Update removals for each checkpoint
                    changes += try_join_all(removals.iter().map(|(checkpoint, removals)| {
                        // We want to remove from the database all name records that were removed in the checkpoint
                        // but only if the checkpoint is newer than the last time the name record was updated.
                        diesel::delete(table)
                            .filter(
                                field_id
                                    .eq_any(removals)
                                    .and(last_checkpoint_updated.le(*checkpoint as i64)),
                            )
                            .execute(conn)
                    }))
                    .await?
                    .iter()
                    .sum::<usize>();

                    Ok(changes)
                }
                .scope_boxed()
            })
            .await?)
    }
}

impl Processor for DomainHandler {
    const NAME: &'static str = "Domain";
    type Value = SuinsCheckpointData;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        let data = checkpoint.transactions.iter().try_fold(
            SuinsIndexerCheckpoint::default(),
            |mut results, tx| {
                // loop through all the transactions in the checkpoint
                // Since the transactions are sequenced inside the checkpoint, we can safely assume
                // that we have the latest data for each name record in the end of the loop.
                // Add all name record changes to the name_records HashMap.
                // Remove any removals that got re-created.
                self.parse_record_changes(&mut results, &tx.output_objects)?;

                // Gather all removals from the transaction,
                // and delete any name records from the name_records if it got deleted.
                self.parse_record_deletions(&mut results, tx);
                Ok::<_, anyhow::Error>(results)
            },
        )?;

        // Convert our name_records & wrappers into a list of updates for the DB.
        Ok(vec![SuinsCheckpointData {
            updates: data.prepare_db_updates(checkpoint.checkpoint_summary.sequence_number),
            removals: data
                .removals
                .into_iter()
                .map(|id| id.to_hex_uncompressed())
                .collect(),
            checkpoint: checkpoint.checkpoint_summary.sequence_number,
        }])
    }
}
