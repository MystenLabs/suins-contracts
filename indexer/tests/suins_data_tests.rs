use chrono::NaiveDateTime;
use fastcrypto::hash::{HashFunction, Sha256};
use insta::assert_json_snapshot;
use move_core_types::language_storage::StructTag;
use serde_json::Value;
use sqlx::{Column, PgPool, Row, ValueRef};
use std::fs;
use std::path::Path;
use std::str::FromStr;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_indexer_alt_framework::store::Store;
use sui_pg_db::temp::TempDb;
use sui_pg_db::Connection;
use sui_pg_db::Db;
use sui_pg_db::DbArgs;
use sui_storage::blob::Blob;
use sui_types::base_types::SuiAddress;
use sui_types::full_checkpoint_content::CheckpointData;
use suins_indexer::handlers::domain_handler::DomainHandler;
use suins_indexer::MIGRATIONS;

const TEST_REGISTRY_TABLE_ID: &str =
    "0xb120c0d55432630fce61f7854795a3463deb6e3b443cc4ae72e1282073ff56e4";
const TEST_NAME_RECORD_TYPE: &str = "0x2::dynamic_field::Field<0x22fa05f21b1ad71442491220bb9338f7b7095fe35000ef88d5400d28523bdd93::domain::Domain,0x22fa05f21b1ad71442491220bb9338f7b7095fe35000ef88d5400d28523bdd93::name_record::NameRecord>";
const TEST_SUBDOMAIN_REGISTRATION_TYPE: &str = "0x22fa05f21b1ad71442491220bb9338f7b7095fe35000ef88d5400d28523bdd93::subdomain_registration::SubDomainRegistration";

/// For our test policy, we have a few checkpoints that contain some data additions, deletions, replacements
///
/// Checkpoint 22279187: Adds 3 different names (1 SLD, 1 leaf, 1 node). Deletes none.
/// Checkpoint 22279365: Removes 1 leaf name. Adds 1 leaf name.
/// Checkpoint 22279496: Replaces the name added on `22279365` (new.test.sui) by removing it and then adding it as a node name.
/// Checkpoint 22279944: Adds `remove.test.sui`.
/// Checkpoint 22280030: Adds `remove.test.sui` as a replacement (the previous one expired!).
///                      [This was only simulated using a dummy contract and cannot happen in realistic scenarios.]
///
///
#[tokio::test]
async fn process_22279187_checkpoint() -> Result<(), anyhow::Error> {
    let handler = get_test_handler();
    data_test("process_22279187_checkpoint", handler, ["domains"]).await?;
    Ok(())
}

#[tokio::test]
async fn process_22279365_checkpoint() -> Result<(), anyhow::Error> {
    let handler = get_test_handler();
    data_test("process_22279365_checkpoint", handler, ["domains"]).await?;
    Ok(())
}

#[tokio::test]
async fn process_22279496_checkpoint() -> Result<(), anyhow::Error> {
    let handler = get_test_handler();
    data_test("process_22279496_checkpoint", handler, ["domains"]).await?;
    Ok(())
}

#[tokio::test]
async fn process_22279944_checkpoint() -> Result<(), anyhow::Error> {
    let handler = get_test_handler();
    data_test("process_22279944_checkpoint", handler, ["domains"]).await?;
    Ok(())
}

#[tokio::test]
async fn process_22280030_checkpoint() -> Result<(), anyhow::Error> {
    let handler = get_test_handler();
    data_test("process_22280030_checkpoint", handler, ["domains"]).await?;
    Ok(())
}

fn get_test_handler() -> DomainHandler {
    let registry = SuiAddress::from_str(TEST_REGISTRY_TABLE_ID).unwrap();
    let subdomain = StructTag::from_str(TEST_SUBDOMAIN_REGISTRATION_TYPE).unwrap();
    let name_record = StructTag::from_str(TEST_NAME_RECORD_TYPE).unwrap();
    DomainHandler::new(registry, subdomain, name_record)
}

async fn data_test<H, I>(
    test_name: &str,
    handler: H,
    tables_to_check: I,
) -> Result<(), anyhow::Error>
where
    I: IntoIterator<Item = &'static str>,
    H: Handler + Processor,
    for<'a> H::Store: Store<Connection<'a> = Connection<'a>>,
{
    // Set up the temporary database
    let temp_db = TempDb::new()?;
    let url = temp_db.database().url();
    let db = Arc::new(Db::for_write(url.clone(), DbArgs::default()).await?);
    db.run_migrations(Some(&MIGRATIONS)).await?;
    let mut conn = db.connect().await?;

    // Test setup based on provided test_name
    let test_path = Path::new("tests/checkpoints").join(test_name);
    let checkpoints = get_checkpoints_in_folder(&test_path)?;

    // Run pipeline for each checkpoint
    for checkpoint in checkpoints {
        run_pipeline(&handler, &checkpoint, &mut conn).await?;
    }

    // Check results by comparing database tables with snapshots
    for table in tables_to_check {
        let rows = read_table(&table, &url.to_string()).await?;
        assert_json_snapshot!(format!("{test_name}__{table}"), rows);
    }
    Ok(())
}

async fn run_pipeline<'c, T: Handler + Processor, P: AsRef<Path>>(
    handler: &T,
    path: P,
    conn: &mut Connection<'c>,
) -> Result<(), anyhow::Error>
where
    T::Store: Store<Connection<'c> = Connection<'c>>,
{
    let bytes = fs::read(path)?;
    let cp = Blob::from_bytes::<CheckpointData>(&bytes)?;
    let result = handler.process(&Arc::new(cp))?;
    T::commit(&result, conn).await?;
    Ok(())
}

/// Read the entire table from database as json value.
/// note: bytea values will be hashed to reduce output size.
async fn read_table(table_name: &str, db_url: &str) -> Result<Vec<Value>, anyhow::Error> {
    let pool = PgPool::connect(db_url).await?;
    let rows = sqlx::query(&format!("SELECT * FROM {table_name}"))
        .fetch_all(&pool)
        .await?;

    // To json
    Ok(rows
        .iter()
        .map(|row| {
            let mut obj = serde_json::Map::new();

            for column in row.columns() {
                let column_name = column.name();

                let value = if let Ok(v) = row.try_get::<String, _>(column_name) {
                    Value::String(v)
                } else if let Ok(v) = row.try_get::<i32, _>(column_name) {
                    Value::String(v.to_string())
                } else if let Ok(v) = row.try_get::<i64, _>(column_name) {
                    Value::String(v.to_string())
                } else if let Ok(v) = row.try_get::<bool, _>(column_name) {
                    Value::Bool(v)
                } else if let Ok(v) = row.try_get::<Value, _>(column_name) {
                    v
                } else if let Ok(v) = row.try_get::<Vec<u8>, _>(column_name) {
                    // hash bytea contents
                    let mut hash_function = Sha256::default();
                    hash_function.update(v);
                    let digest2 = hash_function.finalize();
                    Value::String(digest2.to_string())
                } else if let Ok(v) = row.try_get::<NaiveDateTime, _>(column_name) {
                    Value::String(v.to_string())
                } else if let Ok(true) = row.try_get_raw(column_name).map(|v| v.is_null()) {
                    Value::Null
                } else {
                    panic!(
                        "Cannot parse DB value to json, type: {:?}, column: {column_name}",
                        row.try_get_raw(column_name)
                            .map(|v| v.type_info().to_string())
                    )
                };
                obj.insert(column_name.to_string(), value);
            }

            Value::Object(obj)
        })
        .collect())
}

fn get_checkpoints_in_folder(folder: &Path) -> Result<Vec<String>, anyhow::Error> {
    let mut files = Vec::new();

    // Read the directory
    for entry in fs::read_dir(folder)? {
        let entry = entry?;
        let path = entry.path();

        // Check if it's a file and ends with ".chk"
        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("chk") {
            files.push(path.display().to_string());
        }
    }

    Ok(files)
}
