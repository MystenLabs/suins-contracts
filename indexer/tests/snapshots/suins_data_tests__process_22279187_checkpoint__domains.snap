---
source: indexer/tests/suins_data_tests.rs
expression: rows
---
[
  {
    "name": "leaf.test.sui",
    "parent": "test.sui",
    "expiration_timestamp_ms": 0,
    "nft_id": "0xa4891f3754b203ef230a5e2a08822c835c808eab71e2bc6ca33a73cec9728376",
    "field_id": "0x1f79de658fd87e7746ab672f018a06d1c7b4b9045a97323c140c34e6f072733e",
    "target_address": "0x0000000000000000000000000000000000000000000000000000000000000002",
    "data": {
      "contents": []
    },
    "last_checkpoint_updated": 22279187,
    "subdomain_wrapper_id": null
  },
  {
    "name": "test.sui",
    "parent": "sui",
    "expiration_timestamp_ms": 1737739567137,
    "nft_id": "0xa4891f3754b203ef230a5e2a08822c835c808eab71e2bc6ca33a73cec9728376",
    "field_id": "0x68dab6c226642aaf11f410b1784538ec2b7a836479742f853ea58ef136d16be3",
    "target_address": "0xfe09cf0b3d77678b99250572624bf74fe3b12af915c5db95f0ed5d755612eb68",
    "data": {
      "contents": []
    },
    "last_checkpoint_updated": 22279187,
    "subdomain_wrapper_id": null
  },
  {
    "name": "node.test.sui",
    "parent": "test.sui",
    "expiration_timestamp_ms": 1706213544456,
    "nft_id": "0x1c62ec0f5f4bdf9104748bd3f5211d8fe56df5a6540d57af49fd6631ee8178e3",
    "field_id": "0xe88240e3b30cd39bdf112cfec33820bd9a87e84a403ca2d0c9be0da80029080e",
    "target_address": null,
    "data": {
      "contents": []
    },
    "last_checkpoint_updated": 22279187,
    "subdomain_wrapper_id": "0x5fa94f7596779f6816cbca58d25aba4e8df094d428feea304f2a151be6f95422"
  }
]
